packages/decoder/lib/decoders.ts

```diff
@@ -133,6 +144,146 @@ export class ProjectDecoder {
     debug("done with allocation");
   }
 
+  /**
+   * **This function is asynchronous.**
+   *
+   * Adds compilations to the decoder after it has started.  Note it is
+   * only presently possible to do this with a `ProjectDecoder` and not
+   * with the other decoder classes.
+   *
+   * @param compilations The compilations to be added.  Take care that these
+   * have IDs distinct from those the decoder already has.
+   */
+  public async addCompilations(
+    compilations: Compilations.Compilation[]
+  ): Promise<void> {
+    //first: make sure we're not adding a compilation with an existing ID
+    const existingIds = new Set(
+      this.compilations.map(compilation => compilation.id)
+    );
+    //we use a find() rather than a some() so that we can put the ID in the error
+    const conflictingCompilation = compilations.find(compilation =>
+      existingIds.has(compilation.id)
+    );
+    if (conflictingCompilation !== undefined) {
+      throw new RepeatCompilationIdError(conflictingCompilation.id);
+    }
+    //also: check for repeats among the ones we're adding
+    for (let i = 0; i < compilations.length; i++) {
+      for (let j = i + 1; j < compilations.length; j++) {
```

Comment:
Oh, they're expected to be in order! Can we assume the client breaks this guarantee?

Reply:
They're not expected to be in order.  Why do you say they are?